File: examples/python/plane_wave_3D.py
# To run this script you must install eiko with the optional example dependencies.
# This is done using "pip install eiko[examples]".

import torch
import sys
import torch
import torch.nn.functional as F

# Check if the optional dependencies are available
try:
    import matplotlib.pyplot as plt
    import matplotlib.animation as animation
    from matplotlib.patches import Polygon
except ImportError:
    print("\n[Eiko] Error: This script requires 'matplotlib' for visualization.", file=sys.stderr)
    sys.exit(1)

try:
    from scipy.signal.windows import tukey
except ImportError:
    print("\n[Eiko] Error: This script requires 'scipy' to generate the Tukey window.", file=sys.stderr)
    sys.exit(1)

try:
    from skimage import measure
    HAS_SKIMAGE = True
except ImportError:
    HAS_SKIMAGE = False
    print("[Eiko] Warning: 'scikit-image' not found. Contours and overlays will be disabled.", file=sys.stderr)

import numpy as np
from scipy.signal.windows import tukey
from eiko import eiko, animate_eikonal

# =========================================================================
# 1. Setup 3D Domain and Transducer Aperture
# =========================================================================
canvas_height = int(1080 / 4)
canvas_width  = int(1920 / 4)
canvas_depth  = canvas_width
msfm = False 
steering_angle_deg = 20
theta = np.deg2rad(steering_angle_deg)

device = 'cuda'

aperture_y = 0
aperture_x = slice(int(canvas_width * 0.2), int(canvas_width * 0.8))
aperture_z = slice(int(canvas_depth * 0.2), int(canvas_depth * 0.8))

# Apodization windows
tw_x = tukey(len(range(*aperture_x.indices(canvas_width))))
tw_z = tukey(len(range(*aperture_z.indices(canvas_depth))))
apod_2d = torch.tensor(np.outer(tw_z, tw_x), device=device, dtype=torch.float32)

# Firing delays
delays_x = torch.arange(canvas_width, device=device).float() * np.sin(theta)
delays_x -= delays_x.min()

u_init = torch.full((canvas_height, canvas_width, canvas_depth), float('inf'), device=device)
v_init = torch.zeros((canvas_height, canvas_width, canvas_depth), device=device)

# Assign to aperture
u_init[aperture_y, aperture_x, aperture_z] = delays_x[aperture_x]
v_init[aperture_y, aperture_x, aperture_z] = apod_2d.T 

# =========================================================================
# 2. 3D Plane Wave in a Homogeneous Medium
# =========================================================================
f_homo = torch.ones((canvas_height, canvas_width, canvas_depth), device=device)
u_homo, v_homo = eiko(u_init, f_homo, v_init=v_init, msfm=msfm)
animate_eikonal(u_homo, v_homo, render_mode='slice', title=f'Plane Wave ({steering_angle_deg}° Steered)')


print(f"U min/max: {u_homo.min():.2f} / {u_homo.max():.2f}")
print(f"V min/max: {v_homo.min():.2f} / {v_homo.max():.2f}")

# Check if U is still mostly Inf
print(f"Fraction of valid U: {torch.isfinite(u_homo).float().mean():.2%}")

# =========================================================================
# 3. 3D Plane Wave in an Inhomogeneous Medium (Aberrated)
# =========================================================================
z_grid, y_grid, x_grid = torch.meshgrid(
    torch.arange(canvas_depth, device=device), 
    torch.arange(canvas_height, device=device), 
    torch.arange(canvas_width, device=device), indexing='ij'
)

center_x, center_y, center_z = canvas_width/2, canvas_height*0.35, canvas_depth/2
radius = canvas_width * 0.12
sphere_mask = ((x_grid - center_x)**2 + (y_grid - center_y)**2 + (z_grid - center_z)**2) <= radius**2

f_inhomo = torch.ones_like(x_grid, dtype=torch.float32)
f_inhomo[sphere_mask] = 0.8

u_uncorrected, v_uncorrected = eiko(u_init, f_inhomo, v_init=v_init, msfm=msfm)
animate_eikonal(u_uncorrected, v_uncorrected, render_mode='slice', outline=sphere_mask, 
                title=f'Aberrated Plane Wave ({steering_angle_deg}° Steered)')

# =========================================================================
# 4. 3D Aberration Correction via Time-Reversal
# =========================================================================
# 4a. Back-propagation
u_init_bw = torch.full((canvas_height, canvas_width, canvas_depth), float('inf'), device=device)
u_init_bw[-1, :, :] = torch.arange(canvas_width, device=device).float() * np.sin(theta)

u_bw = eiko(torch.flip(u_init_bw, [0]), torch.flip(f_inhomo, [0]), msfm=msfm)
u_bw = torch.flip(u_bw, [0])

# 4b. Extract delays
arrival_times = u_bw[aperture_y, aperture_x, aperture_z]
delays_corrected = arrival_times.max() - arrival_times

# 4c. Re-initialize
u_init_corrected = torch.full((canvas_height, canvas_width, canvas_depth), float('inf'), device=device)
u_init_corrected[aperture_y, aperture_x, aperture_z] = delays_corrected

# 4d. Forward pass
u_corrected, v_corrected = eiko(u_init_corrected, f_inhomo, v_init=v_init, msfm=msfm)

animate_eikonal(u_corrected, v_corrected, render_mode='slice', outline=sphere_mask, 
                title=f'3D Aberration Corrected ({steering_angle_deg}° Steered)')

File: examples/matlab/plane_wave_3D.m
%% Example: 3D Volumetric Plane-Wave Steering and Aberration Correction
% Simulates a 2D matrix transducer array emitting a 3D steered plane wave.
% Demonstrates propagation in a homogeneous medium, an aberrated medium, 
% and corrects the 3D wavefront using time-reversal.

%% 1. Setup 3D Domain and Transducer Aperture
canvasHeight = 1080/4;
canvasWidth  = 1920/4;
canvasDepth  = canvasWidth; % Make the volume a square block in X-Z

% Global solver settings
msfm = false; % MSFM can be computationally heavy in 3D, disabled by default
steering_angle_deg = 20; 
theta = deg2rad(steering_angle_deg);

% Define the 2D aperture at the top of the canvas (y = 1)
aperture_y = 1;
aperture_x = floor(canvasWidth*0.2) : ceil(canvasWidth*0.8);
aperture_z = floor(canvasDepth*0.2) : ceil(canvasDepth*0.8);

% Create 2D apodization window (v_init) using the outer product of two Tukey windows
tw_x = tukeywin(numel(aperture_x));
tw_z = tukeywin(numel(aperture_z));
apod_2d = tw_z * tw_x'; % 2D weights for the matrix array

% Calculate 1D firing delays for steering in the X direction
delays_x = aperture_x * sin(theta);
delays_x = delays_x - min(delays_x); % Shift so first element fires at t=0

% Apply delays and apodization to the 3D Initial Condition Matrices
u_init = inf(canvasHeight, canvasWidth, canvasDepth, 'single');
v_init = zeros(canvasHeight, canvasWidth, canvasDepth, 'single');

for k = 1:numel(aperture_z)
    z_idx = aperture_z(k);
    u_init(aperture_y, aperture_x, z_idx) = delays_x;
    v_init(aperture_y, aperture_x, z_idx) = apod_2d(k, :);
end

%% 2. 3D Plane Wave in a Homogeneous Medium
fprintf('Simulating 3D plane wave in a homogeneous medium...\n');
f_homo = ones(canvasHeight, canvasWidth, canvasDepth, 'single'); % c=1

[u_homo, v_homo] = eiko(u_init, f_homo, 'v_init', v_init, 'msfm', msfm);

animate_eikonal(u_homo, v_homo, 'RenderMode', 'slice', 'Style', 'real', ...
    'Title', sprintf('Plane Wave (%d° Steered)', steering_angle_deg));

%% 3. 3D Plane Wave in an Inhomogeneous Medium (Aberrated)
fprintf('Simulating 3D plane wave in an aberrated medium...\n');

% Create a 3D spherical anomaly in the center of the volume
[X, Y, Z] = meshgrid(1:canvasWidth, 1:canvasHeight, 1:canvasDepth);
center_x = canvasWidth / 2;
center_y = canvasHeight * 0.35;
center_z = canvasDepth / 2;
radius = canvasWidth * 0.12;

sphere_mask = ((X - center_x).^2 + (Y - center_y).^2 + (Z - center_z).^2) <= radius^2;

% Modify the slowness map (e.g., 20% faster sound speed inside the sphere)
f_inhomo = ones(canvasHeight, canvasWidth, canvasDepth, 'single');
f_inhomo(sphere_mask) = 0.8; 

% Run forward solver with the exact same initial uncorrected delays
[u_uncorrected, v_uncorrected] = eiko(u_init, f_inhomo, 'v_init', v_init, 'msfm', msfm);

animate_eikonal(u_uncorrected, v_uncorrected, 'RenderMode', 'slice', 'Outline', sphere_mask, 'Style', 'real', ...
    'Title', sprintf('Aberrated Plane Wave (%d° Steered)', steering_angle_deg));

%% 4. 3D Aberration Correction via Time-Reversal
fprintf('Performing 3D aberration correction (time-reversal)...\n');

% 4a. Propagate a wave from the bottom face (target) back to the transducer
u_init_bw = inf(canvasHeight, canvasWidth, canvasDepth, 'single');
for z_idx = 1:canvasDepth
    u_init_bw(end, :, z_idx) = -(1:canvasWidth) * sin(theta);
end

u_bw = eiko(flipud(u_init_bw), flipud(f_inhomo), 'msfm', msfm);
u_bw = flipud(u_bw);

% 4b. Extract the highly complex 2D arrival times at the matrix array
arrival_times = u_bw(aperture_y, aperture_x, aperture_z);

% Use the maximum value to invert the wavefront shape
delays_corrected = max(arrival_times(:)) - arrival_times;

% 4c. Re-initialize the forward propagation with the 2D corrected delays
u_init_corrected = inf(canvasHeight, canvasWidth, canvasDepth, 'single');
for k = 1:numel(aperture_z)
    z_idx = aperture_z(k);
    u_init_corrected(aperture_y, aperture_x, z_idx) = delays_corrected(1, :, k);
end

% 4d. Run the corrected forward simulation
[u_corrected, v_corrected] = eiko(u_init_corrected, f_inhomo, 'v_init', v_init, 'msfm', msfm);

animate_eikonal(u_corrected, v_corrected, 'RenderMode', 'slice', 'Outline', sphere_mask, 'Style', 'real', ...
    'Title', sprintf('Aberration Corrected Plane Wave (%d° Steered)', steering_angle_deg));